BuildSourceImage.sh

```diff
@@ -1,80 +1,1072 @@
+    # TODO don't always assume +gzip
+    layer_dgsts="$(jq '.layers[] | select(.mediaType == "application/vnd.oci.image.layer.v1.tar+gzip") | .digest' "${image_dir}"/blobs/"${mnfst_dgst/:/\/}" | tr -d \")"
+    ret=$?
+    if [ ${ret} -ne 0 ] ; then
+        return ${ret}
+    fi
+
+    _mkdir_p "${unpack_dir}/rootfs"
+    for dgst in ${layer_dgsts} ; do
+        path="${image_dir}/blobs/${dgst/:/\/}"
+        tmp_file=$(_mktemp)
+        zcat "${path}" | _tar -t > "$tmp_file"
+
+        # look for '.wh.' entries. They must be removed from the rootfs
+        # _before_ extracting the archive, then the .wh. entries themselves
+        # need to not remain afterwards
+        grep '\.wh\.' "${tmp_file}" | while read -r wh_path ; do
+            # if `some/path/.wh.foo` then `rm -rf `${unpack_dir}/some/path/foo`
+            # if `some/path/.wh..wh..opq` then `rm -rf `${unpack_dir}/some/path/*`
+            if [ "$(basename "${wh_path}")" == ".wh..wh..opq" ] ; then
+                _rm_rf "${unpack_dir}/rootfs/$(dirname "${wh_path}")/*"
+            elif basename "${wh_path}" | grep -qe '^\.wh\.' ; then
+                name=$(basename "${wh_path}" | sed -e 's/^\.wh\.//')
+                _rm_rf "${unpack_dir}/rootfs/$(dirname "${wh_path}")/${name}"
+            fi
+        done
+
+        _info "[unpacking] layer ${dgst}"
+        # unpack layer to rootfs (without whiteouts)
+        zcat "${path}" | _tar --restrict --no-xattr --no-acls --no-selinux --exclude='*.wh.*' -x -C "${unpack_dir}/rootfs"
+        ret=$?
+        if [ ${ret} -ne 0 ] ; then
+            return ${ret}
+        fi
+
+        # some of the directories get unpacked as 0555, so removing them gives an EPERM
+        find "${unpack_dir}" -type d -exec chmod 0755 "{}" \;
+    done
+}
 
 #
-# If the caller specified a extra directory,
-# add it to the CONTEXT DIR in SRC IMAGE
+# unpack using umoci
 #
-if [ ! -z "${EXTRA_SRC_DIR}" ]; then
-    buildah add ${SRC_CTR} ${EXTRA_SRC_DIR} /EXTRA
-    buildah config --created-by "/bin/sh -c #(nop) ADD file:$(cd ${EXTRA_SRC_DIR}; tar cf - . | sha256sum -| cut -f1 -d' ') in /CONTEXT" ${SRC_CTR}
-    export IMG=$(buildah commit --omit-timestamp --rm ${EXTRA_SRC_CTR})
-    export SRC_CTR=$(buildah from ${IMG})
-fi
+unpack_img_umoci() {
+    local image_dir="${1}"
+    local unpack_dir="${2}"
 
-# Cleanup and remove source container
-buildah rm ${SRC_CTR}
+    _debug "unpacking with umoci"
+    # always assume we're not root I reckon
+    umoci unpack --rootless --image "${image_dir}" "${unpack_dir}" >&2
+    ret=$?
+    return $ret
+}
 
 #
-# Add the final name to our image
+# copy an image from one location to another
 #
-buildah tag $IMG $SRC_IMAGE
+push_img() {
+    local src="${1}"
+    local dst="${2}"
+
+    _debug "pushing image ${src} to ${dst}"
+    ## TODO: check for authfile, creds, and whether it's an insecure registry
+    skopeo copy --dest-tls-verify=false "$(ref_prefix "${src}")" "$(ref_prefix "${dst}")" # XXX for demo only
```

Comment:
Reminder: need to revert.

```toml
[registries.insecure]
registries = ["localhost:5000"]
```
^ I have that by default in my `registries.conf` to ease testing.

Reply:
yes this needs to be removed, but I want auth to be in place too